python/ray/autoscaler/_private/gen2/node_provider.py

```diff
@@ -0,0 +1,758 @@
+            )
+
+    def _stopped_nodes(self, tags):
+
+        filter = {
+            TAG_RAY_CLUSTER_NAME: self.cluster_name,
+            TAG_RAY_NODE_KIND: tags[TAG_RAY_NODE_KIND],
+        }
+
+        nodes = []
+        for node_id in self.nodes_tags:
+            try:
+                node_tags = self.nodes_tags[node_id]
+                if all(item in node_tags.items() for item in filter.items()):
+                    node = self.ibm_vpc_client.get_instance(node_id).result
+                    state = node["status"]
+                    if state in ["stopped", "stopping"]:
+                        nodes.append(node)
+            except Exception as e:
+                cli_logger.warning(node_id)
+                if e.message == "Instance not found":
+                    continue
+                raise e
+        return nodes
+
+    def _create_node(self, base_config, tags):
+        name_tag = tags[TAG_RAY_NODE_NAME]
+        assert (
+            len(name_tag) <= (INSTANCE_NAME_MAX_LEN - INSTANCE_NAME_UUID_LEN - 1)
+        ) and re.match("^[a-z0-9-:-]*$", name_tag), (name_tag, len(name_tag))
+
+        # append instance name with uuid
+        name = "{name_tag}-{uuid}".format(
+            name_tag=name_tag, uuid=uuid4().hex[:INSTANCE_NAME_UUID_LEN]
+        )
+
+        # create instance in vpc
+        instance = self._create_instance(name, base_config)
+
+        # currently create and tag is not an atomic operation
+        with self.lock:
+            self.pending_nodes[instance["id"]] = time.time()
+
+        tags[TAG_RAY_CLUSTER_NAME] = self.cluster_name
+        tags[TAG_RAY_NODE_NAME] = name
+        self.set_node_tags(instance["id"], tags)
+
+        # currently always creating public ip for head node
+        if self._get_node_type(name) == NODE_KIND_HEAD:
+            fip_data = self._create_floating_ip(base_config)
+            self._attach_floating_ip(instance, fip_data)
+
+        return {instance["id"]: instance}
+
+    @log_in_out
+    def create_node(self, base_config, tags, count) -> None:
+        stopped_nodes_dict = {}
+        futures = []
+
+        # Try to reuse previously stopped nodes with compatible configs
+        if self.cache_stopped_nodes:
+            stopped_nodes = self._stopped_nodes(tags)
+            stopped_nodes_ids = [n["id"] for n in stopped_nodes]
+            stopped_nodes_dict = {n["id"]: n for n in stopped_nodes}
+
+            if stopped_nodes:
+                cli_logger.print(
+                    f"Reusing nodes {stopped_nodes_ids}. "
+                    "To disable reuse, set `cache_stopped_nodes: False` "
+                    "under `provider` in the cluster configuration."
+                )
+
+            for node in stopped_nodes:
+                logger.info(f"Starting instance {node['id']}")
+                self.ibm_vpc_client.create_instance_action(node["id"], "start")
+
+            time.sleep(1)
```

Comment:
why sleep? is 1 sec enough to avoid flakyness?

Reply:
Actually, added another time.sleep(5) at the end of the create_node(). While testing those changes encountered race issue in the autoscaler.

Reply:
Once, #28150 resolved the sleep(5) can be removed